lib/runtime/flow-control.js

```
@@ -0,0 +1,136 @@
+    the source catches up to real time, it also keeps track of gaps in the
+    points in order to send ticks.
+
+    The Emitter is responsible for coordinating the draining of queues and
+    sending points / ticks down the flowgraph. It loops over all the sources. If
+    any of the queues are empty, it waits until that source reads some points or
+    updates the record to indicate a gap in the points. Once all of the queues
+    have some points in them, it finds the one with the smallest timestamp and
+    emits a burst of points from that source and continues to loop over them
+    sending points from the one with points earliest in time. Any time the
+    Emitter skips a queue because its earliest timestamp is not the earliest
+    one, then it sends a tick from that source to unblock a downstream merge or
+    join point in the flowgraph. Once a source is at eof and its queue is empty,
+    it is removed from consideration.
+
+    Any time a sink has more outstanding points than a given threshold, it can
+    signal the Emitter to pause, at which point it stops sending from any
+    source. The sink is responsible for resuming the Emitter once enough points
+    have been drained.
+
+*/
+
+'use strict';
+
+
+var _ = require('underscore');
+var JuttleLogger = require('../logger');
+var JuttleMoment = require('./types/juttle-moment');
+
+class FlowControl {
+    constructor(options) {
+        this.options = options || {};
+        this.procs = {};
+        this.timer = null;
+
+        this.emitBatchSize = this.options.emitBatchSize || 10000;
+        this.maxQueueSize = this.options.maxQueueSize || 100000;
+        this.defaultQueueSize = this.options.defaultQueueSize || 10000;
+
+        this.logger_name = 'flow-control';
+        this.logger = JuttleLogger.getLogger(this.logger_name);
+    }
+
+    register(proc) {
+        var pname = proc.params.pname;
+        if (!this.procs[pname]) { this.procs[pname] = proc; }
```

`else` should we throw an error or is it OK?